File: 1501-2000/1563.Stone Game V.py
from functools import lru_cache
class Solution:
    def stoneGameV(self, stoneValue: List[int]) -> int:
        @lru_cache(len(stoneValue) + 1)
        def prefix(i):
            if i == -1:
                return 0
            return prefix(i - 1) + stoneValue[i]
        
        def sum_range(i, j):
            return prefix(j - 1) - prefix(i - 1)

        @lru_cache(None)
        def dp(i, j):
            if i + 1 == j:
                return 0
            res = 0
            for split in range(i + 1, j):
                left_sm = sum_range(i, split)
                right_sm = sum_range(split, j)
                if left_sm >= right_sm:
                    res = max(res, right_sm + dp(split, j))
                if left_sm <= right_sm:
                    res = max(res, left_sm + dp(i, split))
            return res
        return dp(0, len(stoneValue))

File: 1501-2000/1563.Stone Game V.java
class Solution {
    private int[] prefix;
    private int[][] table;
    
    private void initTable(int size) {
        table = new int[size][];
        for (int i = 0; i < size; i++) {
            table[i] = new int[size + 1];
            for (int j = 0; j <= size; j++) {
                table[i][j] = -1;
            }
            table[i][i + 1] = 0;
        }
    }
    private void initPrefix(int[] stoneValue) {
        prefix = new int[stoneValue.length + 1];
        for (int i = 0; i < stoneValue.length; i++) {
            prefix[i + 1] = prefix[i] + stoneValue[i];
        }
    }
    private int sumRange(int i, int j) {
        return prefix[j] - prefix[i];
    }
    private int dp(int i, int j) {
        if (table[i][j] == -1) {
            int res = 0;
            for (int split = i + 1; split < j; split++) {
                int lSum = sumRange(i, split);
                int rSum = sumRange(split, j);
                if (lSum >= rSum) {
                    res = Math.max(res, rSum + dp(split, j));
                }
                if (lSum <= rSum) {
                    res = Math.max(res, lSum + dp(i, split));
                }
            }
            table[i][j] = res;
        }
        return table[i][j];
    }
    public int stoneGameV(int[] stoneValue) {
        initPrefix(stoneValue);
        initTable(stoneValue.length);
        return dp(0, stoneValue.length);
    }
}
